federation/executor.go

```diff
@@ -0,0 +1,323 @@
+	}
+
+	var results []interface{}
+	if !isRoot {
+		root, ok := res.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("did not get back a map from executor, got %v", res)
+		}
+
+		federation, ok := root["__federation"].(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("root did not have a federation map, got %v", res)
+		}
+
+		results, ok = federation[typName].([]interface{})
+		if !ok {
+			return nil, fmt.Errorf("federation map did not have a %s slice, got %v", typName, res)
+		}
+	} else {
+		results = []interface{}{res}
+	}
+
+	return results, nil
+}
+
+type pathFollower struct {
+	targets []map[string]interface{}
+	keys    []interface{}
+}
+
+func (pf *pathFollower) extractTargets(node interface{}, path []PathStep) error {
+	// XXX: encode list flattening in path?
+	if slice, ok := node.([]interface{}); ok {
+		for i, elem := range slice {
+			if err := pf.extractTargets(elem, path); err != nil {
+				return fmt.Errorf("idx %d: %v", i, err)
+			}
+		}
+		return nil
+	}
+
+	if len(path) == 0 {
+		obj, ok := node.(map[string]interface{})
+		if !ok {
+			return fmt.Errorf("not an object: %v", obj)
+		}
+		key, ok := obj["__federation"]
+		if !ok {
+			return fmt.Errorf("missing __federation: %v", obj)
+		}
+		pf.targets = append(pf.targets, obj)
+		pf.keys = append(pf.keys, key)
+		return nil
+	}
+
+	obj, ok := node.(map[string]interface{})
+	if !ok {
+		return nil
+	}
+
+	step := path[0]
+	switch step.Kind {
+	case KindField:
+		next, ok := obj[step.Name]
+		if !ok {
+			return fmt.Errorf("does not have key %s", step.Name)
+		}
+
+		if err := pf.extractTargets(next, path[1:]); err != nil {
+			return fmt.Errorf("elem %s: %v", next, err)
+		}
+
+	case KindType:
+		typ, ok := obj["__typename"].(string)
+		if !ok {
+			return fmt.Errorf("does not have string key __typename")
+		}
+
+		if typ == step.Name {
+			if err := pf.extractTargets(obj, path[1:]); err != nil {
+				return fmt.Errorf("typ %s: %v", typ, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (e *Executor) execute(ctx context.Context, p *Plan, keys []interface{}) ([]interface{}, error) {
+	var res []interface{}
+
+	if p.Service == gatewayCoordinatorServiceName {
+		res = []interface{}{
+			map[string]interface{}{},
+		}
+	} else {
+		var err error
+		// Eexecutes that part of the plan (the subquery) on one of the federated gqlservers
+		res, err = e.runOnService(ctx, p.Service, p.Type, keys, p.Kind, p.SelectionSet)
+
+		if err != nil {
+			return nil, fmt.Errorf("run on service: %v", err)
+		}
+	}
+
+	g, ctx := errgroup.WithContext(ctx)
+
+	var resMu sync.Mutex
+
+	for _, subPlan := range p.After {
+		subPlan := subPlan
+		var pf pathFollower
+		if p.Service != gatewayCoordinatorServiceName {
+			if err := pf.extractTargets(res, subPlan.Path); err != nil {
+				return nil, fmt.Errorf("failed to follow path %v: %v", subPlan.Path, err)
+			}
+		} else {
+			pf.keys = nil
+			pf.targets = []map[string]interface{}{
+				res[0].(map[string]interface{}),
```

Comment:
who guarantees that `res` has at least one element and is a map?  Maybe a comment explaining why this is okay would be helpful